javascript/0450-delete-node-in-a-bst.js

```diff
@@ -0,0 +1,40 @@
+/**
+ * Recursion
+ * h = height of the tree;
+ * Time O(h) | Space O(h)
+ * Definition for a binary tree node.
+ * function TreeNode(val, left, right) {
+ *     this.val = (val===undefined ? 0 : val)
+ *     this.left = (left===undefined ? null : left)
+ *     this.right = (right===undefined ? null : right)
+ * }
+ */
+/**
+ * @param {TreeNode} root
+ * @param {number} key
+ * @return {TreeNode}
+ */
+var deleteNode = function(root, key) {
+    if(!root) return root;
+
+    if(key === root.val) {
+        if(!root.left) return root.right;
+        if(!root.right) return root.left;
+
+        // find the smallest val in right bst
+        let curr = root.right;
+        while(curr.left) {
+            curr = curr.left;
+        }
+        // change the curr value
+        root.val = curr.val;
+
+        root.right = deleteNode(root.right, root.val);
+    } else if(key < root.val) {
```

Comment:
Avoid else
```js
/**
 * DFS - BST
 * Time O(Log N) | Space O(Log N)
 * https://leetcode.com/problems/delete-node-in-a-bst/
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = (root, key) => {
    const isBaseCase = (root === null);
    if (isBaseCase) return null;

    const isTarget = (root.val === key);
    if (isTarget) return updateChildren(root)

    const isGreater = (root.val <= key)
    if (isGreater) root.right = deleteNode(root.right, key);

    const isLess = (key < root.val)
    if (isLess) root.left = deleteNode(root.left, key);

    return root;
};

const updateChildren = (root) => {
    const isLeaf = !(root.left || root.right);
    if (isLeaf) return root.left || root.right;

    let curr = root.right
    while (curr.left) curr = curr.left;
    curr.left = root.left;

    return root.right
}
```

Reply:
Check out this solution.
Submission link for the proposed code: https://leetcode.com/problems/delete-node-in-a-bst/submissions/1388696779/

```javascript
/**
 * Recursion
 * h = height of the tree;
 * Time O(h) | Space O(h)
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
    if(!root) return root;

    if(key === root.val) {
        if(!root.left) return root.right;
        if(!root.right) return root.left;

        // find the smallest val in right bst
        let curr = root.right;
        while(curr.left) {
            curr = curr.left;
        }
        // change the curr value
        root.val = curr.val;

        root.right = deleteNode(root.right, root.val);

        return root;
    } 
    if(key < root.val) {
        root.left = deleteNode(root.left, key);
        return root;
    }
    root.right = deleteNode(root.right, key);
    return root;
};

```

Reply:
@aakhtar3 (tagging to draw attention to the proposed code)